### PR TITLE
fix(acp/codex): fail the prompt on a failed turn; stop repeating agent messages

### PR DESCRIPTION
**Pull request by azmaveth**
Lands the existing `fix/codex-failed-turns-and-duplicate-text` branch (commit 784a59d), brought up to date with master (#32, #34, #35) via a merge commit. Neither fix was on master: master's Codex adapter has no `turn_failure`, `unstreamed_text`, or `last_error` handling and no tests for failed turns or duplicated text.

## Failed turns fail the prompt

codex-cli 0.151 app-server reports transport and quota failures as an `error` notification (`usageLimitExceeded`, a 401 response-stream disconnect, a system error) followed by `turn/completed` with `status: "failed"`. The adapter mapped neither, so the ACP prompt resolved as a successful `end_turn` with empty text and callers could not tell "the model said nothing" from "the model never ran".

`error` notifications are now remembered on the session, and `turn/completed` with a failed or errored turn answers the active prompt with a JSON-RPC error carrying the Codex message, `codexErrorInfo`, and a classified `kind`: `rate_limit_exhausted` (-32029), `unauthenticated` (-32031), `turn_failed` (-32030). The last `error` notification is used when the turn carries no error of its own.

## No duplicated agent messages

`item/completed` for an agent message emitted the full text again after it had been streamed via `item/agentMessage/delta`, so chunk consumers saw it twice. The final chunk now carries only the text not already streamed, and that remainder is folded into the accumulated text so `_meta.ex_mcp.text` matches the stream even without deltas.

## Merge

The only conflict with master was CHANGELOG.md; the two Fixed bullets now sit under Unreleased alongside the mint bump. The Codex adapter file auto-merged cleanly with the #35 namespace and `name/0` changes.

## Validation (OTP 27.3.4 / Elixir 1.17.3, merged tree)

- `mix format --check-formatted`, `mix compile --warnings-as-errors`, `mix credo --strict`, `mix dialyzer`: clean.
- `mix test test/ex_mcp/acp test/ex_mcp/security` under `env -i`: 673 tests, 5 doctests, 0 failures, including the six regression tests from the branch.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_017S1LGhfEJSxn1khq4mnG39

---
_Generated by [Claude Code](https://claude.ai/code/session_017S1LGhfEJSxn1khq4mnG39)_

---

> [!NOTE]
> **Medium Risk**
> Changes prompt completion semantics for Codex failures (callers may now see errors instead of empty success) and streaming chunk boundaries, which can affect error handling and UI that assumed duplicate final chunks.
> 
> **Overview**
> Fixes two Codex app-server integration bugs in the ACP adapter.
> 
> **Failed turns no longer look like empty successes.** When codex-cli reports quota/auth/transport failures via an `error` notification followed by `turn/completed` with `status: "failed"`, the adapter now stores that error on the session and completes the active prompt with a JSON-RPC **error** (including `codexErrorInfo` and a classified `kind`: `rate_limit_exhausted` / `unauthenticated` / `turn_failed`) instead of `stopReason: "end_turn"` with blank text. Failures are checked before the existing empty-text rate-limit heuristic.
> 
> **Agent message chunks are no longer duplicated.** Streaming deltas are tracked per `itemId` in `:streamed_items`; on `item/completed`, only the **remainder** after streamed text is emitted as the final chunk and merged into the prompt accumulator so `_meta.ex_mcp.text` matches what clients saw on the wire.
> 
> CHANGELOG documents both fixes; six regression tests cover failed-turn error codes and multi-message streaming edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2908b81e87af6d042a65b8b90887205fe5a12b0a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lib/ex_mcp/acp/adapters/codex.ex [hunk 1]
```diff
@@ -2208,9 +2218,34 @@ defmodule ExMCP.ACP.Adapters.Codex do
...
+    remainder = unstreamed_text(text, streamed)
+
+    state =
+      if remainder == "" do
+        state
+      else
+        Sessions.update(state, session_id, fn session ->
+          Map.update(session, :accumulated_text, [remainder], &[remainder | &1])
+        end)
+      end
```
[the remark this answers is not in the document]
Comment (azmaveth) [reply lib/ex_mcp/acp/adapters/codex.ex]: Confirmed and fixed in 2908b81. Deltas are now recorded per `itemId` (with a `:current` fallback when app-server omits it), `item/completed` compares against its own item's stream and clears that record, and the per-item map resets with the other prompt accumulators. Added a regression test for a three-message turn (fully streamed, no deltas, partially streamed) and moved the extends-the-stream test onto the delta path.

---
_Generated by [Claude Code](https://claude.ai/code)_
